app/lib/service/entrypoint/_isolate.dart

```diff
@@ -45,83 +46,168 @@ class DebugMessage extends Message {
-  final _isolates = <_Isolate>[];
+  final _groups = <IsolateGroup>[];
 
   IsolateRunner({
     required this.logger,
   });
 
-  Future<void> startIsolates({
+  /// Starts a new isolate group with [count] running instances.
+  @visibleForTesting
+  Future<IsolateGroup> startGroup({
     required String kind,
     required Future<void> Function(EntryMessage message) entryPoint,
     required int count,
     required Duration? deadTimeout,
   }) async {
-    int started = 0;
-
-    Future<void> start() async {
-      if (_closing) return;
-      started++;
-      final id = '$kind isolate #$started';
-      logger.info('About to start $id ...');
-      final isolate = _Isolate(
-        runner: this,
-        logger: logger,
-        id: id,
-        entryPoint: entryPoint,
-      );
-      _isolates.add(isolate);
-      await isolate.init(deadTimeout: deadTimeout);
-      logger.info('$id started.');
-      _lastStarted = clock.now();
-
-      // automatic restart logic
-      unawaited(isolate.done.then((_) async {
-        _isolates.remove(isolate);
-        if (_closing) return;
-        // Restart the isolate after a pause, increasing the pause duration at
-        // each restart.
+    if (_closing) {
+      throw AssertionError('Runner is closed.');
+    }
+    final group = IsolateGroup(
+      runner: this,
+      kind: kind,
+      entryPoint: entryPoint,
+      deadTimeout: deadTimeout,
+    );
+    _groups.add(group);
+    await group.start(count);
+    return group;
+  }
+
+  Future<void> _closeGroups() async {
+    while (_groups.isNotEmpty) {
+      await _groups.removeLast().close();
+    }
+  }
+
+  Future<void> close() async {
+    _closing = true;
+    await _closeGroups();
+  }
+}
+
+/// Starts, monitors, stops or restarts isolates that run the same code.
+///
+/// Once an isolate starts, it is expected to run indefinitely. When it exits,
+/// either by completion or uncaught exception, a new isolate will be started.
+class IsolateGroup {
+  final IsolateRunner runner;
+  final String kind;
+  final Future<void> Function(EntryMessage message) entryPoint;
+  final Duration? deadTimeout;
+  final bool skipWaitBetweenRestarts;
+
+  int started = 0;
+  final _isolates = <_Isolate>[];
+  bool get _closing => runner._closing;
+  Logger get logger => runner.logger;
+
+  /// The duration while internal errors won't cause isolates to restart.
+  var _restartProtectionOffset = Duration.zero;
+  var _lastStarted = clock.now();
+
+  IsolateGroup({
+    required this.runner,
+    required this.kind,
+    required this.entryPoint,
+    required this.deadTimeout,
+    this.skipWaitBetweenRestarts = false,
+  });
+
+  /// Starts [count] new isolates.
+  Future<void> start(int count) async {
+    for (var i = 0; i < count; i++) {
+      await _startOne();
+    }
+  }
+
+  /// Starts [count] new isolates, and after a [wait] duration,
+  /// closes the old ones.
+  Future<void> renew({
+    required int count,
+    required Duration wait,
```

Comment:
Wouldn't it be safer to wait to tell the isolates to shutdown when they are not handling any more search messages...
and then have a timeout for sanity, so that if the timeout expires we hard kill.
For this wouldn't need a configurable timeout, just a hardcoded sanity value and a shout into logs.

Reply:
Yeah, that's probably better to do. I though that we can skip the complexity by giving it enough time for surely completing the requests (and the rest should already timed out by the calling HTTP handlers).